modules/ds18b20/ds18b20.c

```diff
@@ -0,0 +1,433 @@
+    /* Select ROM number */
+    OneWire_SelectWithPointer(OneWireStruct, ROM);
+    /* Read scratchpad command by OneWireStruct protocol */
+    OneWire_WriteByte(OneWireStruct, ONEWIRE_CMD_RSCRATCHPAD);
+    
+    /* Get data */
+    for (i = 0; i < DS18B20_DATA_LEN; i++) {
+        /* Read byte by byte */
+        data[i] = OneWire_ReadByte(OneWireStruct);
+    }
+    
+    /* Calculate CRC */
+    crc = OneWire_CRC8(data, DS18B20_DATA_CRC_BYTE);
+    
+    /* Check if CRC is ok */
+    if (crc != data[DS18B20_DATA_CRC_BYTE]) {
+        /* CRC invalid */
+        return DS18B20_CRC_INVALID;
+    }
+    
+    /* First two bytes of scratchpad are temperature values */
+    temperature = data[DS18B20_DATA_LSB] | (data[DS18B20_DATA_MSB] << 8);
+
+    /* Reset line */
+    OneWire_Reset(OneWireStruct);
+    
+    /* Check if temperature is negative */
+    if (temperature & DS18B20_TEMP_SIGN_BITMASK) {
+        /* Two's complement, temperature is negative */
+        temperature = ~temperature + 1;
+        minus = 1;
+    }
+
+    /* Get sensor resolution */
+    /* Extract resolution bits with &, Ignore unused bits in the configuration register 
+    which are reserved for internal use by the device. Add 9 to convert R0 and R1 -> 9-12 for 
+    bits in resolution */
+    resolution = ((data[DS18B20_CONFIG_REGISTER_BYTE] & DS18B20_CONFIG_REGISTER_R0_R1_BITMASK) 
+        >> DS18B20_CONFIG_REGISTER_RESERVED_BITS) + 9;
+    
+    /* Store temperature integer digits and decimal digits */
+    /* Ignore all but 1 signed bit */
+    digit = temperature >> 4;
+    /* extract signed bit and 3 digit temp bits from MSB, put in appropriate power */
+    digit |= ((temperature >> 8) & 0x7) << 4;
+    
+    /* Store decimal digits */
+    /* Resolution may be either 9, 10, 11, or 12 bits.
+    If the DS18B20  is  configured for  12-bit resolution, all bits in the temperature
+    register will contain valid  data. For 11-bit resolution, bit 0 is undefined. 
+    For  10-bit resolution, bits 1 and 0 are undefined, and for 9-bit resolution  
+    bits  2,  1,  and  0  are  undefined. Shift as needed to extract 1 step of 
+    the decimal temperature then multiply by decimal steps per resolution */
+    switch (resolution) {
+        case 9: {
+            decimal = (temperature >> 3) & 0x01;
+            decimal *= (float)DS18B20_DECIMAL_STEPS_9BIT;
+        } break;
+        case 10: {
+            decimal = (temperature >> 2) & 0x03;
+            decimal *= (float)DS18B20_DECIMAL_STEPS_10BIT;
+        } break;
+        case 11: {
+            decimal = (temperature >> 1) & 0x07;
+            decimal *= (float)DS18B20_DECIMAL_STEPS_11BIT;
+        } break;
+        case 12: {
+            decimal = temperature & 0x0F;
+            decimal *= (float)DS18B20_DECIMAL_STEPS_12BIT;
+        } break;
+        default: {
+            decimal = 0xFF;
+            digit = 0;
+        }
+    }
+    
+    /* Check for negative part */
+    decimal = digit + decimal;

```

Comment:
This comment doesn't appear correct for line 155 but rather 156.  

I think a better comment is needed for what is happens in line 155.